src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -4979,5 +4794,1571 @@ await foreach (var o in new Enumerable())
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: @"123Disposed");
+        }
+
+        [Fact]
+        public void TestGetAsyncEnumeratorPatternViaExtensionsWithTaskLikeTypeMoveNext()
+        {
+            string source = @"
+using System;
+using System.Threading.Tasks;
+public class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public ValueTask<bool> MoveNextAsync() => new ValueTask<bool>(Current++ != 3);
+    }
+}
+public static class Extensions
+{
+    public static C.Enumerator GetAsyncEnumerator(this C self) => new C.Enumerator();
+}";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_IAsyncEnumerable }, options: TestOptions.DebugExe, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "123");
+        }
+
+        [Fact]
+        public void TestWithObsoletePatternMethodsViaExtension()
+        {
+            string source = @"
+using System;
+using System.Threading.Tasks;
+public class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        [Obsolete]
+        public int Current { get; private set; }
+        [Obsolete]
+        public Task<bool> MoveNextAsync() => Task.FromResult(Current++ != 3);
+        [Obsolete]
+        public Task DisposeAsync() { Console.Write(""Disposed""); return Task.CompletedTask; }
+    }
+}
+[Obsolete]
+public static class Extensions
+{
+    [Obsolete]
+    public static C.Enumerator GetAsyncEnumerator(this C self) => new C.Enumerator();
+}";
+            var comp = CreateCompilationWithMscorlib46(source, options: TestOptions.DebugExe, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (8,15): warning CS0612: 'Extensions.GetAsyncEnumerator(C)' is obsolete
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("Extensions.GetAsyncEnumerator(C)").WithLocation(8, 15),
+                // (8,15): warning CS0612: 'C.Enumerator.MoveNextAsync()' is obsolete
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.Enumerator.MoveNextAsync()").WithLocation(8, 15),
+                // (8,15): warning CS0612: 'C.Enumerator.Current' is obsolete
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.Enumerator.Current").WithLocation(8, 15));
```

Comment:
Why don't we see a diagnostic for `DisposeAsync`?

Reply:
I don't know, but this is an existing bug (applies even to [ordinary Dispose methods](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGAzAATakDChA3oaQ+VlvY3QY56QGbQwCGYABakAFADd+UUgEtZAO1LyYAdyqiAlBtZdS7XbsoR5AZwgAbGADoA6lBkAXGKJkaA3Dq4BfTz46MSUhMBSwATJlIAUXkAVwBbGCh+B2hSEFIASQARGRMABwgTfmBLT30DQJl5ByoYqFhqvVIAcxgHN1I8+0knILaOvwNyMmAIC1IAWQhxGAA5GAQHTVIAXgA+UgcoGJgPf10AbQB5YDNLJ1EAIkuAGlJEqGh0rZ2NAF1PTkCMFFIc/MKzg0NEGjFBDEC0XiiWSqQA4m0oQkkikoMt1pshI81LFzOY9l4gA))

Do you want me to fix this here or create a new issue?

Reply:
That's... annoying. I'll file a bug, fixing this would be a breaking change :( (warning where we do not today).

Reply:
There's an existing bug: https://github.com/dotnet/roslyn/issues/30257